core/src/main/java/org/elasticsearch/index/query/RandomSampleQuery.java

```diff
@@ -0,0 +1,318 @@
+                    };
+                }
+
+            }
+        };
+    }
+
+    @Override
+    public String toString(String field) {
+        return "RandomSample[p=" + this.p + "](*:*)";
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        RandomSampleQuery other = (RandomSampleQuery)o;
+        return Objects.equals(this.p, other.p) &&
+            Objects.equals(this.seed, other.seed);
+    }
+
+    @Override
+    public int hashCode() {
+        return classHash();
+    }
+
+    /**
+     * A DocIDSetIter that samples on each document.  Empirically, this tends to
+     * be more accurate when the segment is small.  It may also be faster
+     * when the probability is high, since many docs are collected and gap approximation
+     * uses more expensive math
+     */
+    private static class RandomSamplingDocIdSetIterator extends DocIdSetIterator {
+        int doc = -1;
+        final int maxDoc;
+        final Random rng;
+        final double p;
+
+        RandomSamplingDocIdSetIterator(int maxDoc, Random rng, double p) {
+            this.maxDoc = maxDoc;
+            this.rng = rng;
+            this.p = p;
+        }
+
+        @Override
+        public int docID() {
+            return doc;
+        }
+
+        @Override
+        public int nextDoc() throws IOException {
+            return advance(doc + 1);
+        }
+
+        @Override
+        public int advance(int target) throws IOException {
+            doc = target;
+            while (doc < maxDoc) {
+                if (rng.nextDouble() <= p) {
+                    return doc;
+                }
+                doc = doc + 1;
+            }
+            return NO_MORE_DOCS;
+        }
+
+        @Override
+        public long cost() {
+            return maxDoc;
+        }
+    }
+
+    /**
+     * A DocIDSetIter that approximates the gaps between sampled documents, and advances
+     * according to the gap.  This is more efficient, especially for low probabilities,
+     * because it can skip by many documents entirely.
+     */
+    private static class RandomGapSamplingDocIdSetIterator extends DocIdSetIterator {
+        final int maxDoc;
+        final Random rng;
+        final double p;
+        int doc = -1;
+
+        RandomGapSamplingDocIdSetIterator(int maxDoc, Random rng, double p) {
+            this.maxDoc = maxDoc;
+            this.rng = rng;
+            this.p = p;
+        }
+
+        @Override
+        public int docID() {
+            return doc;
+        }
+
+        @Override
+        public int nextDoc() throws IOException {
+            return advance(doc);
+        }
+
+        @Override
+        public int advance(int target) throws IOException {
+            // Keep approximating gaps until we hit or surpass the target
+            while (doc <= target) {
+                doc += getGap(rng);
+            }
```

Comment:
I'm slightly worried this could be slow if the percentage of docs to match is rather high and `target-doc` is large?